bamt/utils/MathUtils.py

```diff
@@ -117,4 +122,107 @@ def component(data, columns, method):
                 biggest_p = p
                 comp_biggest = i
         n = comp_biggest
-    return n
\ No newline at end of file
+    return n
+
+
+def get_n_nearest(data, columns, corr=False, number_close=5):
+    """Returns N nearest neighbors for every column of dataframe, added into list
+
+    Args:
+        data (DataFrame): Proximity matrix
+        columns (list): df.columns.tolist()
+        corr (bool, optional): _description_. Defaults to False.
+        number_close (int, optional): Number of nearest neighbors. Defaults to 5.
+
+    Returns:
+        groups
+    """
+    groups = []
+    for c in columns:
+        if corr:
+            close_ind = data[c].sort_values(ascending=False).index.tolist()
+        else:
+            close_ind = data[c].sort_values().index.tolist()
+        groups.append(close_ind[0:number_close + 1])
+
+    return groups
+
+
+def get_proximity_matrix(df, df_coded, proximity_metric) -> pd.DataFrame:
+    """Returns matrix of proximity matrix of the dataframe, dataframe must be coded first if it contains
+                                                                                                    categorical data
+
+    Args:
+        df (DataFrame): data
+        df_coded (DataFrame): same data, but coded
+        proximity_metric (str): 'MI' or 'corr'
+
+    Returns:
+        df_distance: mutual information matrix
+    """
+
+    df_distance = pd.DataFrame(data=np.zeros(
+        (len(df.columns), len(df.columns))), columns=df.columns)
+    df_distance.index = df.columns
+
+    if proximity_metric == 'MI':
+        for c1 in df.columns:
+            for c2 in df.columns:
+                dist = mutual_info_score(
+                    df_coded[c1].values, df_coded[c2].values)
+                df_distance.loc[c1, c2] = dist
+
+    elif proximity_metric == 'corr':
+        df_distance = df_coded.corr(method='pearson')
+
+    return df_distance
+
+
+def get_brave_matrix(df, proximity_matrix, n_nearest=5) -> pd.DataFrame:
```

Comment:
It seems that here the source dataframe is used only to obtain a list of features (columns), so can it be done so that this list is initially passed to the function input? Why pull the entire dataframe?

Reply:
Fixed it, thanks